importer: support arrays of UDT in some cases

yuzefovich opened the pull request:
This commit adjusts the import code to support IMPORT INTO tables with columns typed as arrays of UDTs. This required a couple of minor changes:
- propagating the `SemaCtx` into `ParseDatumStringAs`. Previously, we would always create a fresh one, but now all callers of this method have been adjusted to provide the one they have in scope. This allows us to parse expressions of the form `'Monday'::db.sc.enum_name` where `'Monday'` is a member of an enum.
- implement `importTypeResolver.ResolveType` in the "happy" case. This is used to resolve casts from above to a concrete type.

Note that `ResolveType` implementation is incomplete in the general case. In particular, whenever an import job is created, we _might_ have a set of types available (it appears that this is the case when we're importing into one table - for example, we have logic for importing the whole pgdump, there we can have multiple tables, and set of types won't be available). Whenever we do have a set of types, we can simplify the type resolution to simply match on the name of the type.

However, if we happen to have a table that uses two UDTs with the same name but different schemas, this simplistic resolution won't work, so we still return an error in this case.

Fixes: #112100.

Release note (sql change): CockroachDB now supports IMPORT INTO a table that has columns typed as arrays of user-defined types (like enums). Tables that uses multiple user-defined types with the same name but different schemas are still unsupported.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/115146)
<!-- Reviewable:end -->

yuzefovich commented:
@fqazi @Xiang-Gu I'd appreciate a quick look at implementation of `importTypeResolver.ResolveType`. I think we can by-pass lots of the resolution steps whenever we have `[]*descpb.TypeDescriptor` provided to us by the import spec. It seems to me that whenever we're importing into an existing table, we will have that provided as an argument, and in such cases we can simply match on the non-qualified type name. This doesn't work in general case, but it might be good enough to make things work in the common case (at least it makes the simply reproduction in the commit message work).

yuzefovich committed:
importer: support arrays of UDT in some cases

This commit adjusts the import code to support IMPORT INTO tables with
columns typed as arrays of UDTs. This required a couple of minor
changes:
- propagating the `SemaCtx` into `ParseDatumStringAs`. Previously, we
would always create a fresh one, but now all callers of this method have
been adjusted to provide the one they have in scope. This allows us to
parse expressions of the form `'Monday'::db.sc.enum_name` where
`'Monday'` is a member of an enum.
- implement `importTypeResolver.ResolveType` in the "happy" case. This
is used to resolve casts from above to a concrete type.

Note that `ResolveType` implementation is incomplete in the general
case. In particular, whenever an import job is created, we _might_ have
a set of types available (it appears that this is the case when we're
importing into one table - for example, we have logic for importing the
whole pgdump, there we can have multiple tables, and set of types won't
be available). Whenever we do have a set of types, we can simplify the
type resolution to simply match on the name of the type.

However, if we happen to have a table that uses two UDTs with the same
name but different schemas, this simplistic resolution won't work, so we
still return an error in this case.

Release note (sql change): CockroachDB now supports IMPORT INTO a table
that has columns typed as arrays of user-defined types (like enums).
Tables that uses multiple user-defined types with the same name but
different schemas are still unsupported.

yuzefovich reviewed (commented):
I polished this change for the happy case, so it's RFAL. It comes with some limitations, but I'd guess it should work in the common cases.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-:-Nl0Dm6z5lrP_Y-RG1Dy:b2txm50)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @DrewKimball, @fqazi, and @Xiang-Gu)

___
*[`-- commits` line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl0DUX6-rFOMpy4Anbd):*
@adityamaru can you double check me on this? In particular, if I'm reading the code right, in `importPlanHook` whenever we're importing into a single table, we're populating `[]ImportDetails_Type` which we then use when creating `importTypeResolver`. That set of type descriptors contains all types in the table. Does this sound right?


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Have some fairly minor comments for the resolver logic

> Reviewed 3 of 12 files at r1, 11 of 13 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-:-Nl0VXfw9dOAsrimTTjt:b-u0jllt)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @DrewKimball, @Xiang-Gu, and @yuzefovich)

___
*[`pkg/sql/importer/import_type_resolver.go` line 62 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl0U9lYAOo-pHfl-Nrx:-Nl0U9lYAOo-pHfl-Nry:b-l5ktsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a15e24adc7f253e56a9dc019ee25cf8344e4f6a/pkg/sql/importer/import_type_resolver.go#L62)):*
> ```Go
> 	var ok bool
> 	if descs, ok = i.typeNameToDesc[name.Parts[0]]; !ok || len(descs) == 0 {
> 		return nil, errors.Newf("type descriptor could not be resolved for type name %s", name.Parts[0])
> ```

Lets go with: NewUndefinedTypeError
___
*[`pkg/sql/importer/import_type_resolver.go` line 65 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl0UfOn7upEdm21trgW:-Nl0UfOn7upEdm21trgX:bexymly) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a15e24adc7f253e56a9dc019ee25cf8344e4f6a/pkg/sql/importer/import_type_resolver.go#L65)):*
> ```Go
> 	}
> 	if len(descs) > 1 {
> 		return nil, errors.Newf("tables with multiple user-defined types with the same name are currently unsupported: %s", name.Parts[0])
> ```

Lets do a pgerror with the pgcode: pgcode.FeatureNotSupported here


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @DrewKimball, @fqazi, and @Xiang-Gu)

___
*[`pkg/sql/importer/import_type_resolver.go` line 62 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl0U9lYAOo-pHfl-Nrx:-Nl0XHtW3WwybNyNZVrI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a15e24adc7f253e56a9dc019ee25cf8344e4f6a/pkg/sql/importer/import_type_resolver.go#L62)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Lets go with: NewUndefinedTypeError
</blockquote></details>

Done.
___
*[`pkg/sql/importer/import_type_resolver.go` line 65 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl0UfOn7upEdm21trgW:-Nl0XI89CIT7_DpPbKoL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a15e24adc7f253e56a9dc019ee25cf8344e4f6a/pkg/sql/importer/import_type_resolver.go#L65)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Lets do a pgerror with the pgcode: pgcode.FeatureNotSupported here
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> Reviewed 2 of 12 files at r1, 8 of 13 files at r2, 1 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @fqazi, @Xiang-Gu, and @yuzefovich)

___
*[`pkg/sql/importer/import_type_resolver.go` line 65 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl5PlX250rMbkg2NmAO:-Nl5PlX250rMbkg2NmAP:b-5s6c8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f7e524efdb2b780c56163b0db8f803b531a75d/pkg/sql/importer/import_type_resolver.go#L65)):*
> ```Go
> 	var descs []*descpb.TypeDescriptor
> 	var ok bool
> 	if descs, ok = i.typeNameToDesc[name.Parts[0]]; !ok || len(descs) == 0 {
> ```

Could this resolve the wrong type if the descriptor referenced one type with the same name, but the imported expression referenced another? As in, table references `db1.typ`, and the import casts an expression to `db2.typ`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @DrewKimball, @fqazi, and @Xiang-Gu)

___
*[`pkg/sql/importer/import_type_resolver.go` line 65 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl5PlX250rMbkg2NmAO:-Nl5bMkX6Yv_enDcYoRW:b-4pwnrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f7e524efdb2b780c56163b0db8f803b531a75d/pkg/sql/importer/import_type_resolver.go#L65)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could this resolve the wrong type if the descriptor referenced one type with the same name, but the imported expression referenced another? As in, table references `db1.typ`, and the import casts an expression to `db2.typ`.
</blockquote></details>

That's a good question. I think it is possible with the current implementation. For example, with the following:
```sql
CREATE DATABASE db1;  
USE db1;  
CREATE TYPE weekday AS ENUM('Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday');  
CREATE DATABASE db2;  
USE db2;  
CREATE TYPE weekday AS ENUM('Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday');  
USE db1;  
CREATE TABLE db1_shifts (employee STRING, days weekday[]);  
INSERT INTO db1_shifts VALUES ('John', ARRAY['Monday', 'Wednesday', 'Friday']);  
INSERT INTO db1_shifts VALUES ('Bob', ARRAY['Tuesday', 'Thursday']);  
  
EXPORT INTO CSV 'nodelocal://1/export1/' FROM SELECT * FROM db1_shifts;  
  
USE db2;  
CREATE TABLE db2_shifts (employee STRING, days weekday[]);  
IMPORT INTO db2_shifts CSV DATA ('nodelocal://1/export1/export*-n*.0.csv');  
SELECT * FROM db2_shifts;
```
we have csv files of the form:
```
John,"ARRAY['Monday':::db1.public.weekday,'Wednesday':::db1.public.weekday,'Friday':::db1.public.weekday]"
Bob,"ARRAY['Tuesday':::db1.public.weekday,'Thursday':::db1.public.weekday]"
```
yet the snippet above succeeds. In other words, if two user-defined types happen to have the same logical representation (at least for those values that are present in the data set being imported), then we will silently read UDT values of one type as UDT values of another.

This seems like an edge case, and I'm not sure what's the expected behavior should be. Probably the IMPORT should fail, right?

I'll spend some more time trying to see whether I can fully implement ideas that Faizan and Xiang mentioned on slack to make this resolution logic work in all cases.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @DrewKimball, @fqazi, and @Xiang-Gu)

___
*[`pkg/sql/importer/import_type_resolver.go` line 65 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl5PlX250rMbkg2NmAO:-Nl5w9YuBUCIku0og2Ya:bi1qurv) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f7e524efdb2b780c56163b0db8f803b531a75d/pkg/sql/importer/import_type_resolver.go#L65)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

That's a good question. I think it is possible with the current implementation. For example, with the following:
```sql
CREATE DATABASE db1;  
USE db1;  
CREATE TYPE weekday AS ENUM('Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday');  
CREATE DATABASE db2;  
USE db2;  
CREATE TYPE weekday AS ENUM('Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday');  
USE db1;  
CREATE TABLE db1_shifts (employee STRING, days weekday[]);  
INSERT INTO db1_shifts VALUES ('John', ARRAY['Monday', 'Wednesday', 'Friday']);  
INSERT INTO db1_shifts VALUES ('Bob', ARRAY['Tuesday', 'Thursday']);  
  
EXPORT INTO CSV 'nodelocal://1/export1/' FROM SELECT * FROM db1_shifts;  
  
USE db2;  
CREATE TABLE db2_shifts (employee STRING, days weekday[]);  
IMPORT INTO db2_shifts CSV DATA ('nodelocal://1/export1/export*-n*.0.csv');  
SELECT * FROM db2_shifts;
```
we have csv files of the form:
```
John,"ARRAY['Monday':::db1.public.weekday,'Wednesday':::db1.public.weekday,'Friday':::db1.public.weekday]"
Bob,"ARRAY['Tuesday':::db1.public.weekday,'Thursday':::db1.public.weekday]"
```
yet the snippet above succeeds. In other words, if two user-defined types happen to have the same logical representation (at least for those values that are present in the data set being imported), then we will silently read UDT values of one type as UDT values of another.

This seems like an edge case, and I'm not sure what's the expected behavior should be. Probably the IMPORT should fail, right?

I'll spend some more time trying to see whether I can fully implement ideas that Faizan and Xiang mentioned on slack to make this resolution logic work in all cases.
</blockquote></details>

Hm, implementing the type resolution proper seems tricky, and I don't think it's worth spending more time on this since it seems like an edge case. I filed #115842 to track this, and I'm actually not sure whether we do want such an import to fail - if we're able to parse strings that were created for `db1.t` type as `db2.t` type, especially with the output of EXPORT (which uses `:::` meaning it's a "type hint"), I don't see a problem with that.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @DrewKimball, @fqazi, and @Xiang-Gu)

___
*[`pkg/sql/importer/import_type_resolver.go` line 65 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl5PlX250rMbkg2NmAO:-NlVoppqC-KJ-8Yv-N2z:bsvq4cr) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f7e524efdb2b780c56163b0db8f803b531a75d/pkg/sql/importer/import_type_resolver.go#L65)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, implementing the type resolution proper seems tricky, and I don't think it's worth spending more time on this since it seems like an edge case. I filed #115842 to track this, and I'm actually not sure whether we do want such an import to fail - if we're able to parse strings that were created for `db1.t` type as `db2.t` type, especially with the output of EXPORT (which uses `:::` meaning it's a "type hint"), I don't see a problem with that.
</blockquote></details>

I think we should merge the change as is (and backport it too since it's quite safe) because it works in vast majority of cases, and even in the edge case like I mentioned above the behavior seems acceptable. @DrewKimball do you agree?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @fqazi, @Xiang-Gu, and @yuzefovich)

___
*[`pkg/sql/importer/import_type_resolver.go` line 65 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl5PlX250rMbkg2NmAO:-NlVsN3X2kOsGIPw7M5X:b-9lere6) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f7e524efdb2b780c56163b0db8f803b531a75d/pkg/sql/importer/import_type_resolver.go#L65)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think we should merge the change as is (and backport it too since it's quite safe) because it works in vast majority of cases, and even in the edge case like I mentioned above the behavior seems acceptable. @DrewKimball do you agree?
</blockquote></details>

I'm leery of introducing incorrect behavior in place of an error, although I do agree it seems unlikely users would hit the edge case. I guess I'd be alright with it if we fix the issue fairly soon?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @fqazi, @Xiang-Gu, and @yuzefovich)

___
*[`pkg/sql/importer/import_type_resolver.go` line 65 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl5PlX250rMbkg2NmAO:-NlW-MOM3Mb_DjmeXTIm:b8bbyub) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f7e524efdb2b780c56163b0db8f803b531a75d/pkg/sql/importer/import_type_resolver.go#L65)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I'm leery of introducing incorrect behavior in place of an error, although I do agree it seems unlikely users would hit the edge case. I guess I'd be alright with it if we fix the issue fairly soon?
</blockquote></details>

How would you feel about introducing unused `[]*descpb.SchemaDescriptor` and `[]*descpb.DatabaseDescriptor` slices to `execinfrapb.ReadImportDataSpec` now to make the fix easier?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Aside from being uncomfortable with the edge case we talked about.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-:-NlW-uk01D3ipR22roGS:b-nn0y0p)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @fqazi, @Xiang-Gu, and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @DrewKimball, @fqazi, and @Xiang-Gu)

___
*[`pkg/sql/importer/import_type_resolver.go` line 65 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl5PlX250rMbkg2NmAO:-NlW-Snx-VktsPuAqSVQ:b9ahzt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f7e524efdb2b780c56163b0db8f803b531a75d/pkg/sql/importer/import_type_resolver.go#L65)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

How would you feel about introducing unused `[]*descpb.SchemaDescriptor` and `[]*descpb.DatabaseDescriptor` slices to `execinfrapb.ReadImportDataSpec` now to make the fix easier?
</blockquote></details>

I doubt we'll fix it any time soon unless a customer comes to us complaining about this. However, I'm not sure that it is incorrect behavior per se - like if a customer creates a CSV file with data referencing `db1.custom_type` but then attempts to import that file into a table that uses `db2.custom_type` (but doesn't use `db1.custom_type`, and the "logical" representation of values of that type are the same), what should be the expected behavior? Would they like for such an import to succeed or result in an error?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @DrewKimball, @fqazi, and @Xiang-Gu)

___
*[`pkg/sql/importer/import_type_resolver.go` line 65 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-Nl5PlX250rMbkg2NmAO:-NlW1gVzEEqkygTkP3L-:brrwr3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/93f7e524efdb2b780c56163b0db8f803b531a75d/pkg/sql/importer/import_type_resolver.go#L65)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I doubt we'll fix it any time soon unless a customer comes to us complaining about this. However, I'm not sure that it is incorrect behavior per se - like if a customer creates a CSV file with data referencing `db1.custom_type` but then attempts to import that file into a table that uses `db2.custom_type` (but doesn't use `db1.custom_type`, and the "logical" representation of values of that type are the same), what should be the expected behavior? Would they like for such an import to succeed or result in an error?
</blockquote></details>

I'd rather not introduce anything unused given that it'll only be used in this particular edge case. There is a more general approach required to make this `importTypeResolver` perform the proper resolution logic, in the absence of provided type descriptors in the import job details. I think we should punt addressing this edge case until we have a customer asking us for the more general solution, and then we'll introduce any changes to the protos that we'll need at that time. However, I suspect it'll be years before someone will complain / ask about this.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115146#-:-NlYzLtLAoClvBN9CVda:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @fqazi, and @Xiang-Gu)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+
